tests/nativeapp/codegen/snowpark/test_extension_function_utils.py

```diff
@@ -92,3 +181,127 @@ def test_get_all_imports():
     )
+
+
+def test_is_function_wellformed():
+    ex_fn = {}
+    assert not ef_utils._is_function_wellformed(ex_fn=ex_fn)  # noqa: SLF001
+
+    ex_fn["func"] = []
+    assert not ef_utils._is_function_wellformed(ex_fn=ex_fn)  # noqa: SLF001
+
+    ex_fn["func"] = " "
+    assert not ef_utils._is_function_wellformed(ex_fn=ex_fn)  # noqa: SLF001
+
+    ex_fn["func"] = "dummy"
+    assert ef_utils._is_function_wellformed(ex_fn=ex_fn)  # noqa: SLF001
+
+    ex_fn["func"] = (None, " ")
+    assert not ef_utils._is_function_wellformed(ex_fn=ex_fn)  # noqa: SLF001
+
+    ex_fn["func"] = (None, "dummy")
+    assert ef_utils._is_function_wellformed(ex_fn=ex_fn)  # noqa: SLF001
+
+
+def test_add_defaults_to_extension_function_required_keys():
+    ex_fn = {}
+    some_path = Path("some/path")
+    with pytest.raises(ef_utils.MalformedExtensionFunctionError) as err:
+        ef_utils.sanitize_extension_function_data(ex_fn=ex_fn, py_file=some_path)
+    assert err.value.message.__contains__("object_type")
+
+    ex_fn["object_type"] = "function"
+    with pytest.raises(ef_utils.MalformedExtensionFunctionError) as err:
+        ef_utils.sanitize_extension_function_data(ex_fn=ex_fn, py_file=some_path)
+    assert err.value.message.__contains__("object_name")
+
+    ex_fn["object_name"] = "some_name"
+    with pytest.raises(ef_utils.MalformedExtensionFunctionError) as err:
+        ef_utils.sanitize_extension_function_data(ex_fn=ex_fn, py_file=some_path)
+    assert err.value.message.__contains__("return_sql")
+
+    ex_fn["return_sql"] = "returns null"
+    with pytest.raises(ef_utils.MalformedExtensionFunctionError) as err:
+        ef_utils.sanitize_extension_function_data(ex_fn=ex_fn, py_file=some_path)
+    assert err.value.message.__contains__("func")
+
+    # Func possibilities already tested in a test case above this
+
+    ex_fn["func"] = "dummy_func"
+    with pytest.raises(ef_utils.MalformedExtensionFunctionError) as err:
+        ef_utils.sanitize_extension_function_data(ex_fn=ex_fn, py_file=some_path)
+    assert err.value.message.__contains__("raw_imports")
+
+    ex_fn["raw_imports"] = ["some/path"]
+    with pytest.raises(ef_utils.MalformedExtensionFunctionError) as err:
+        ef_utils.sanitize_extension_function_data(ex_fn=ex_fn, py_file=some_path)
+    assert err.value.message.__contains__("schema")
+
+    ex_fn["schema"] = "core"
+    with pytest.raises(ef_utils.MalformedExtensionFunctionError) as err:
+        ef_utils.sanitize_extension_function_data(ex_fn=ex_fn, py_file=some_path)
+    assert err.value.message.__contains__("runtime_version")
+
+    ex_fn["runtime_version"] = "3.8"
+    assert ef_utils.sanitize_extension_function_data(ex_fn=ex_fn, py_file=some_path)
+    # This check is because we cannot stop the function execution beyond required keys, but there are other tests below that go in depth
+
+    assert ex_fn["object_type"] == "FUNCTION"
```

Comment:
nit: dictionaries define equality, so you can just assert the entire dictionary at once. As a bonus, this makes sure there are no extra keys.